Index: clientbase.h
===================================================================
--- clientbase.h	(revision 4023)
+++ clientbase.h	(working copy)
@@ -365,6 +365,7 @@
        * @param ext The extension to register.
        */
       void registerStanzaExtension( StanzaExtension* ext );
+	StanzaExtensionFactory * stanzaExtensionFactory( ) { return m_seFactory; }
 
       /**
        * Removes the given StanzaExtension type from the StanzaExtensionFactory.
Index: sihandler.h
===================================================================
--- sihandler.h	(revision 4023)
+++ sihandler.h	(working copy)
@@ -52,7 +52,7 @@
        * @param sid The stream ID.
        * @param si The request's complete SI.
        */
-      virtual void handleSIRequestResult( const JID& from, const std::string& sid,
+      virtual void handleSIRequestResult( const JID& from, const JID& to, const std::string& sid,
                                           const SIManager::SI& si ) = 0;
 
       /**
Index: siprofileft.h
===================================================================
--- siprofileft.h	(revision 4023)
+++ siprofileft.h	(working copy)
@@ -196,7 +196,7 @@
        * @return The requested stream's ID (SID). Empty if conditions above (file name, size)
        * are not met.
        */
-      const std::string requestFT( const JID& to, const std::string& name, long size,
+      const std::string requestFT( const JID& to, const JID& from, const std::string& name, long size,
                                    const std::string& hash = EmptyString,
                                    const std::string& desc = EmptyString,
                                    const std::string& date = EmptyString,
@@ -211,7 +211,7 @@
        * @param type The desired stream type to use for this file transfer. Defaults to
        * SOCKS5 Bytestream. You should not use @c FTTypeAll here.
        */
-      void acceptFT( const JID& to, const std::string& sid,
+      void acceptFT( const JID& to, const JID& from, const std::string& sid,
                      StreamType type = FTTypeS5B );
 
       /**
@@ -282,11 +282,11 @@
         { if( m_socks5Manager ) m_socks5Manager->removeSOCKS5BytestreamServer(); }
 
       // reimplemented from SIProfileHandler
-      virtual void handleSIRequest( const JID& from, const std::string& id,
+      virtual void handleSIRequest( const JID& from, const JID& to, const std::string& id,
                                     const SIManager::SI& si );
 
       // reimplemented from SIHandler
-      virtual void handleSIRequestResult( const JID& from, const std::string& sid,
+      virtual void handleSIRequestResult( const JID& from, const JID& to, const std::string& sid,
                                           const SIManager::SI& si );
 
       // reimplemented from SIHandler
Index: examples/ft_send.cpp
===================================================================
--- examples/ft_send.cpp	(revision 4023)
+++ examples/ft_send.cpp	(working copy)
@@ -124,7 +124,7 @@
     virtual void onConnect()
     {
       printf( "connected!!!\n" );
-      f->requestFT( m_to, m_file, m_size );
+      f->requestFT( m_to, j->jid(), m_file, m_size );
     }
 
     virtual void onDisconnect( ConnectionError e )
@@ -152,7 +152,7 @@
       printf("log: level: %d, area: %d, %s\n", level, area, message.c_str() );
     }
 
-    virtual void handleFTRequest( const JID& from, const std::string& sid,
+    virtual void handleFTRequest( const JID& from, const JID& to, const std::string& sid,
                                   const std::string& name, long size, const std::string& hash,
                                   const std::string& date, const std::string& mimetype,
                                   const std::string& desc, int /*stypes*/, long /*offset*/, long /*length*/ )
@@ -161,7 +161,7 @@
               "desc: %s\n",
               from.full().c_str(), name.c_str(), size, sid.c_str(), hash.c_str(), date.c_str(),
               mimetype.c_str(), desc.c_str() );
-      f->acceptFT( from, sid, SIProfileFT::FTTypeS5B );
+      f->acceptFT( from, to, sid, SIProfileFT::FTTypeS5B );
     }
 
 //     virtual void handleFTRequestResult( const JID& /*from*/, const std::string& /*sid*/ )
@@ -188,7 +188,7 @@
       }
     }
 
-    virtual const std::string handleOOBRequestResult( const JID& /*from*/, const std::string& /*sid*/ )
+    virtual const std::string handleOOBRequestResult( const JID& /*from*/, const JID& /*to*/, const std::string& /*sid*/ )
     {
       return std::string();
     };
Index: examples/ft_recv.cpp
===================================================================
--- examples/ft_recv.cpp	(revision 4023)
+++ examples/ft_recv.cpp	(working copy)
@@ -99,7 +99,7 @@
       printf("log: level: %d, area: %d, %s\n", level, area, message.c_str() );
     }
 
-    virtual void handleFTRequest( const JID& from, const std::string& sid,
+    virtual void handleFTRequest( const JID& from, const JID& to, const std::string& sid,
                                   const std::string& name, long size, const std::string& hash,
                                   const std::string& date, const std::string& mimetype,
                                   const std::string& desc, int /*stypes*/, long /*offset*/, long /*length*/ )
@@ -108,7 +108,7 @@
               "desc: %s\n",
               from.full().c_str(), name.c_str(), size, sid.c_str(), hash.c_str(), date.c_str(),
               mimetype.c_str(), desc.c_str() );
-      f->acceptFT( from, sid, SIProfileFT::FTTypeIBB );
+      f->acceptFT( from, to, sid, SIProfileFT::FTTypeIBB );
     }
 
 //     virtual void handleFTRequestResult( const JID& /*from*/, const std::string& /*sid*/ )
@@ -134,7 +134,7 @@
       }
     }
 
-    virtual const std::string handleOOBRequestResult( const JID& /*from*/, const std::string& /*sid*/ )
+    virtual const std::string handleOOBRequestResult( const JID& /*from*/, const JID& /*to*/, const std::string& /*sid*/ )
     {
       return std::string();
     };
Index: presence.h
===================================================================
--- presence.h	(revision 4023)
+++ presence.h	(working copy)
@@ -138,7 +138,7 @@
 
       // reimplemented from Stanza
       virtual Tag* tag() const;
-
+      Presence( Tag* tag );
     private:
 #ifdef PRESENCE_TEST
     public:
@@ -147,8 +147,8 @@
        * Creates a Presence request from the given Tag. The original Tag will be ripped off.
        * @param tag The Tag to parse.
        */
-      Presence( Tag* tag );
 
+
       PresenceType m_subtype;
       StringMap* m_stati;
       std::string m_status;
Index: simanager.h
===================================================================
--- simanager.h	(revision 4023)
+++ simanager.h	(working copy)
@@ -158,7 +158,7 @@
        * @note The SIManager claims ownership of the Tags supplied to this function, and will
        * delete them after use.
        */
-      const std::string requestSI( SIHandler* sih, const JID& to, const std::string& profile, Tag* child1,
+      const std::string requestSI( SIHandler* sih, const JID& to, const JID& from, const std::string& profile, Tag* child1,
                                    Tag* child2 = 0, const std::string& mimetype = "binary/octet-stream" );
 
       /**
@@ -172,7 +172,7 @@
        * @note The SIManager claims ownership of the Tags supplied to this function, and will
        * delete them after use.
        */
-      void acceptSI( const JID& to, const std::string& id, Tag* child1, Tag* child2 = 0 );
+      void acceptSI( const JID& to, const JID& from, const std::string& id, Tag* child1, Tag* child2 = 0 );
 
       /**
        * Call this function to decline an SI request previously announced by means of
Index: disco.cpp
===================================================================
--- disco.cpp	(revision 4023)
+++ disco.cpp	(working copy)
@@ -265,6 +265,7 @@
       case IQ::Get:
       {
         IQ re( IQ::Result, iq.from(), iq.id() );
+        re.setFrom(iq.to());
 
         const SoftwareVersion* sv = iq.findExtension<SoftwareVersion>( ExtVersion );
         if( sv )
Index: siprofileft.cpp
===================================================================
--- siprofileft.cpp	(revision 4023)
+++ siprofileft.cpp	(working copy)
@@ -60,7 +60,7 @@
       delete m_socks5Manager;
   }
 
-  const std::string SIProfileFT::requestFT( const JID& to, const std::string& name, long size,
+  const std::string SIProfileFT::requestFT( const JID& to, const JID& from, const std::string& name, long size,
                                             const std::string& hash, const std::string& desc,
                                             const std::string& date, const std::string& mimetype,
                                             int streamTypes )
@@ -93,10 +93,10 @@
     dff->setOptions( sm );
     feature->addChild( df.tag() );
 
-    return m_manager->requestSI( this, to, XMLNS_SI_FT, file, feature, mimetype );
+    return m_manager->requestSI( this, to, from, XMLNS_SI_FT, file, feature, mimetype );
   }
 
-  void SIProfileFT::acceptFT( const JID& to, const std::string& sid, StreamType type )
+  void SIProfileFT::acceptFT( const JID& to, const JID& from, const std::string& sid, StreamType type )
   {
     if( !m_manager )
       return;
@@ -119,7 +119,7 @@
         if( m_handler )
         {
           InBandBytestream* ibb = new InBandBytestream( m_parent, m_parent->logInstance(), to,
-                                                        m_parent->jid(), sid );
+                                                        from, sid );
           m_handler->handleFTBytestream( ibb );
         }
         break;
@@ -131,7 +131,7 @@
     df.addField( dff );
     feature->addChild( df.tag() );
 
-    m_manager->acceptSI( to, id, 0, feature );
+    m_manager->acceptSI( to, from, id, 0, feature );
   }
 
   void SIProfileFT::declineFT( const JID& to, const std::string& sid, SIManager::SIError reason,
@@ -166,7 +166,7 @@
       m_socks5Manager->addStreamHost( jid, host, port );
   }
 
-  void SIProfileFT::handleSIRequest( const JID& from, const std::string& id,
+  void SIProfileFT::handleSIRequest( const JID& from, const JID& to, const std::string& id,
                                      const SIManager::SI& si )
   {
     if( si.profile() != XMLNS_SI_FT || !si.tag1() )
@@ -214,7 +214,7 @@
 
       const std::string& sid = si.id();
       m_id2sid[sid] = id;
-      m_handler->handleFTRequest( from, sid, si.tag1()->findAttribute( "name" ),
+      m_handler->handleFTRequest( from, to, sid, si.tag1()->findAttribute( "name" ),
                                   atol( si.tag1()->findAttribute( "size" ).c_str() ),
                                         si.tag1()->findAttribute( "hash" ),
                                             si.tag1()->findAttribute( "date" ),
@@ -223,7 +223,7 @@
     }
   }
 
-  void SIProfileFT::handleSIRequestResult( const JID& from, const std::string& sid,
+  void SIProfileFT::handleSIRequestResult( const JID& from, const JID& to, const std::string& sid,
                                            const SIManager::SI& si )
   {
     if( si.tag2() )
@@ -236,24 +236,25 @@
         if( m_socks5Manager && dff->value() == XMLNS_BYTESTREAMS )
         {
           // check return value:
-          m_socks5Manager->requestSOCKS5Bytestream( from, SOCKS5BytestreamManager::S5BTCP, sid );
+          m_socks5Manager->requestSOCKS5Bytestream( from, to, SOCKS5BytestreamManager::S5BTCP, sid );
         }
         else if( m_handler )
         {
           if( dff->value() == XMLNS_IBB )
           {
             InBandBytestream* ibb = new InBandBytestream( m_parent, m_parent->logInstance(),
-                                                          m_parent->jid(), from, sid );
+                                                          to, from, sid );
 
             m_handler->handleFTBytestream( ibb );
           }
           else if( dff->value() == XMLNS_IQ_OOB )
           {
-            const std::string& url = m_handler->handleOOBRequestResult( from, sid );
+            const std::string& url = m_handler->handleOOBRequestResult( from, to, sid );
             if( !url.empty() )
             {
               const std::string& id = m_parent->getID();
               IQ iq( IQ::Set, from, id );
+			  iq.setFrom(to);
               iq.addExtension( new OOB( url, EmptyString, true ) );
               m_parent->send( iq, this, OOBSent );
             }
Index: tests/simanager/simanager_test.cpp
===================================================================
--- tests/simanager/simanager_test.cpp	(revision 4023)
+++ tests/simanager/simanager_test.cpp	(working copy)
@@ -48,10 +48,10 @@
       void registerStanzaExtension( StanzaExtension* ext );
       void removeStanzaExtension( int ext );
       void removeIDHandler( IqHandler* ) {}
-      virtual void handleSIRequestResult( const JID& from, const std::string& sid,
+      virtual void handleSIRequestResult( const JID& from, const JID& to, const std::string& sid,
                                           const SIManager::SI& si );
       virtual void handleSIRequestError( const IQ& iq, const std::string& /*sid*/ );
-      virtual void handleSIRequest( const JID& from, const std::string& id, const SIManager::SI& si );
+      virtual void handleSIRequest( const JID& from, const JID& to, const std::string& id, const SIManager::SI& si );
       void setTest( int test );
       bool ok();
     private:
@@ -85,10 +85,10 @@
   void ClientBase::registerIqHandler( IqHandler*, int ) {}
   void ClientBase::registerStanzaExtension( StanzaExtension* se ) { delete se; }
   void ClientBase::removeStanzaExtension( int ) {}
-  void ClientBase::handleSIRequestResult( const JID& /*from*/, const std::string& /*sid*/,
+  void ClientBase::handleSIRequestResult( const JID& /*from*/, const JID& /*to*/, const std::string& /*sid*/,
                                           const SIManager::SI& /*si*/ ) {}
   void ClientBase::handleSIRequestError( const IQ& /*iq*/, const std::string& /*sid*/ ) {}
-  void ClientBase::handleSIRequest( const JID& /*from*/, const std::string& /*id*/,
+  void ClientBase::handleSIRequest( const JID& /*from*/, const JID& /*to*/, const std::string& /*id*/,
                                     const SIManager::SI& /*si*/ ) {}
   void ClientBase::setTest( int test ) { m_test = test; }
   bool ClientBase::ok() { bool t = m_ok; m_ok = false; return t; }
@@ -119,7 +119,7 @@
   // -------
   name = "request si";
   cb->setTest( 1 );
-  sim->requestSI( cb, to, g_profile, t1->clone(), t2->clone() );
+  sim->requestSI( cb, to, gloox::JID("abcd@def.gh/test"), g_profile, t1->clone(), t2->clone() );
   if( !cb->ok() )
   {
     ++fail;
Index: siprofilehandler.h
===================================================================
--- siprofilehandler.h	(revision 4023)
+++ siprofilehandler.h	(working copy)
@@ -52,7 +52,7 @@
        * SIManager::acceptSI() or SIManager::declineSI().
        * @param si The request's complete SI.
        */
-      virtual void handleSIRequest( const JID& from, const std::string& id, const SIManager::SI& si ) = 0;
+      virtual void handleSIRequest( const JID& from, const JID& to, const std::string& id, const SIManager::SI& si ) = 0;
 
   };
 
Index: siprofilefthandler.h
===================================================================
--- siprofilefthandler.h	(revision 4023)
+++ siprofilefthandler.h	(working copy)
@@ -61,7 +61,7 @@
        * @param length The number of bytes to send, starting from the given offset. A value of -1
        * indicates that the entire file is to be transmitted (taking the offset into account).
        */
-      virtual void handleFTRequest( const JID& from, const std::string& sid,
+      virtual void handleFTRequest( const JID& from, const JID& to, const std::string& sid,
                                     const std::string& name, long size, const std::string& hash,
                                     const std::string& date, const std::string& mimetype,
                                     const std::string& desc, int stypes, long offset, long length ) = 0;
@@ -96,7 +96,7 @@
        * @param sid The stream's ID.
        * @return The file's URL.
        */
-      virtual const std::string handleOOBRequestResult( const JID& from, const std::string& sid ) = 0;
+      virtual const std::string handleOOBRequestResult( const JID& from, const JID& to, const std::string& sid ) = 0;
 
   };
 
Index: socks5bytestreammanager.cpp
===================================================================
--- socks5bytestreammanager.cpp	(revision 4023)
+++ socks5bytestreammanager.cpp	(working copy)
@@ -175,7 +175,7 @@
     m_hosts.push_back( sh );
   }
 
-  bool SOCKS5BytestreamManager::requestSOCKS5Bytestream( const JID& to, S5BMode mode,
+  bool SOCKS5BytestreamManager::requestSOCKS5Bytestream( const JID& to, const JID& from, S5BMode mode,
                                                          const std::string& sid )
   {
     if( !m_parent )
@@ -192,12 +192,13 @@
     const std::string& id = m_parent->getID();
     IQ iq( IQ::Set, to, id );
     iq.addExtension( new Query( msid, mode, m_hosts ) );
+	iq.setFrom(from);
 
     if( m_server )
     {
       SHA sha;
       sha.feed( msid );
-      sha.feed( m_parent->jid().full() );
+      sha.feed( from.full() );
       sha.feed( to.full() );
       m_server->registerHash( sha.hex() );
     }
@@ -206,6 +207,7 @@
     asi.sHosts = m_hosts;
     asi.id = id;
     asi.from = to;
+	asi.to = from;
     asi.incoming = false;
     m_asyncTrackMap[msid] = asi;
 
@@ -229,6 +231,7 @@
     if( item.incoming )
     {
       iq = new IQ( IQ::Result, item.from.full(), item.id );
+	  iq->setFrom(item.to.full());
       if( success )
         iq->addExtension( new Query( jid, sid, false ) );
       else
@@ -274,6 +277,7 @@
         asi.sHosts = q->hosts();
         asi.id = iq.id();
         asi.from = iq.from();
+		asi.to = iq.to();
         asi.incoming = true;
         m_asyncTrackMap[sid] = asi;
         m_socks5BytestreamHandler->handleIncomingBytestreamRequest( sid, iq.from() );
@@ -330,7 +334,7 @@
 
     SOCKS5Bytestream* s5b = new SOCKS5Bytestream( this, m_parent->connectionImpl()->newInstance(),
                                                   m_parent->logInstance(),
-                                                  (*it).second.from, m_parent->jid(), sid );
+                                                  (*it).second.from, (*it).second.to, sid );
     s5b->setStreamHosts( (*it).second.sHosts );
     m_s5bMap[sid] = s5b;
     m_socks5BytestreamHandler->handleIncomingBytestream( s5b );
@@ -414,18 +418,18 @@
                 {
                   SHA sha;
                   sha.feed( (*it).second );
-                  sha.feed( m_parent->jid().full() );
+                  sha.feed( iq.to().full() );
                   sha.feed( iq.from().full() );
                   s5b = new SOCKS5Bytestream( this, m_server->getConnection( sha.hex() ),
                                               m_parent->logInstance(),
-                                                  m_parent->jid(), iq.from(),
+                                                  iq.to(), iq.from(),
                                                       (*it).second );
                 }
                 else
                 {
                   s5b = new SOCKS5Bytestream( this, m_parent->connectionImpl()->newInstance(),
                                               m_parent->logInstance(),
-                                                  m_parent->jid(), iq.from(),
+                                                  iq.to(), iq.from(),
                                                       (*it).second );
                   s5b->setStreamHosts( StreamHostList( 1, *sh ) );
                 }
Index: simanager.cpp
===================================================================
--- simanager.cpp	(revision 4023)
+++ simanager.cpp	(working copy)
@@ -109,7 +109,7 @@
     }
   }
 
-  const std::string SIManager::requestSI( SIHandler* sih, const JID& to, const std::string& profile,
+  const std::string SIManager::requestSI( SIHandler* sih, const JID& to, const JID& from, const std::string& profile,
                                           Tag* child1, Tag* child2, const std::string& mimetype )
   {
     if( !m_parent || !sih )
@@ -120,6 +120,7 @@
 
     IQ iq( IQ::Set, to, id );
     iq.addExtension( new SI( child1, child2, id2, mimetype, profile ) );
+	iq.setFrom(from);
 
     TrackStruct t;
     t.sid = id2;
@@ -131,10 +132,11 @@
     return id2;
   }
 
-  void SIManager::acceptSI( const JID& to, const std::string& id, Tag* child1, Tag* child2 )
+  void SIManager::acceptSI( const JID& to, const JID& from, const std::string& id, Tag* child1, Tag* child2 )
   {
     IQ iq( IQ::Result, to, id );
     iq.addExtension( new SI( child1, child2 ) );
+	iq.setFrom(from);
     m_parent->send( iq );
   }
 
@@ -198,7 +200,7 @@
     if( it != m_handlers.end() && (*it).second )
     {
       // FIXME: don't pass si->tag()!
-      (*it).second->handleSIRequest( iq.from(), iq.id(), *si );
+      (*it).second->handleSIRequest( iq.from(), iq.to(), iq.id(), *si );
       return true;
     }
 
@@ -230,7 +232,7 @@
 
             // FIXME: remove above commented code and
             // check corectness of last 3 params!
-            (*it).second.sih->handleSIRequestResult( iq.from(), (*it).second.sid, *si );
+            (*it).second.sih->handleSIRequestResult( iq.from(), iq.to(), (*it).second.sid, *si );
             m_track.erase( it );
           }
         }
Index: clientbase.cpp
===================================================================
--- clientbase.cpp	(revision 4023)
+++ clientbase.cpp	(working copy)
@@ -1249,13 +1249,15 @@
     bool res = false;
 
     // FIXME remove for 1.1
-//     typedef IqHandlerMapXmlns::const_iterator IQciXmlns;
-//     std::pair<IQciXmlns, IQciXmlns> g = m_iqNSHandlers.equal_range( iq.xmlns() );
+//     typedef IqHandlerMapXmlns::const_iterator IQciXmlns
+//     Tag *tag = iq.tag()->xmlns();
+//     std::pair<IQciXmlns, IQciXmlns> g = m_iqNSHandlers.equal_range( tag->xmlns() );
 //     for( IQciXmlns it = g.first; it != g.second; ++it )
 //     {
 //       if( (*it).second->handleIq( iq ) )
 //         res = true;
 //     }
+//     delete tag;
 
     typedef IqHandlerMap::const_iterator IQci;
     const StanzaExtensionList& sel = iq.extensions();
Index: socks5bytestreammanager.h
===================================================================
--- socks5bytestreammanager.h	(revision 4023)
+++ socks5bytestreammanager.h	(working copy)
@@ -106,7 +106,7 @@
        * @b not indicate that the bytestream has been opened. This is announced by means of the
        * BytestreamHandler.
        */
-      bool requestSOCKS5Bytestream( const JID& to, S5BMode mode, const std::string& sid = EmptyString );
+      bool requestSOCKS5Bytestream( const JID& to, const JID& from, S5BMode mode, const std::string& sid = EmptyString );
 
       /**
        * To get rid of a bytestream (i.e., close and delete it), call this function. You
@@ -281,6 +281,7 @@
       struct AsyncS5BItem
       {
         JID from;
+		JID to;
         std::string id;
         StreamHostList sHosts;
         bool incoming;
